File: app/services/order_service.py
import os
import uuid
from decimal import Decimal
from typing import List

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from .. import models, schemas
from . import ticket_service
from ..qr import generate_qr_code_base64

STRIPE_SECRET_KEY = os.getenv("STRIPE_SECRET_KEY", "").strip()

_stripe = None
if STRIPE_SECRET_KEY:
    try:
        import stripe as _stripe_module

        _stripe_module.api_key = STRIPE_SECRET_KEY
        _stripe = _stripe_module
    except ImportError:
        _stripe = None


async def create_order(db: Session, data: schemas.OrderRequest, user: models.User) -> schemas.OrderResponse:
    tickets: List[models.Ticket] = []
    total_amount = Decimal("0")

    for ticket_id in data.ticket_ids:
        ticket = db.query(models.Ticket).filter(models.Ticket.id == ticket_id).first()
        if ticket is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Ingresso não encontrado: {ticket_id}")
        if ticket.status != models.TicketStatus.RESERVED or ticket.reserved_by != user.id:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"Ingresso não está reservado para este usuário: {ticket_id}",
            )
        tickets.append(ticket)
        total_amount += Decimal(ticket.price)

    order = models.Order(user_id=user.id, total_amount=total_amount, status=models.OrderStatus.PENDING)
    db.add(order)
    db.flush()

    for ticket in tickets:
        db.add(models.OrderItem(order_id=order.id, ticket_id=ticket.id, price=ticket.price))

    payment_intent_id = None
    client_secret = None
    demo_mode = True

    if _stripe is not None:
        try:
            intent = _stripe.PaymentIntent.create(
                amount=int(total_amount * 100),
                currency="usd",
                metadata={"orderId": order.id, "userId": user.id},
            )
            payment_intent_id = intent.id
            client_secret = intent.client_secret
            order.stripe_payment_intent = payment_intent_id
            demo_mode = False
        except Exception as exc:  # pragma: no cover - only hit with real Stripe keys
            db.rollback()
            raise HTTPException(
                status_code=status.HTTP_502_BAD_GATEWAY, detail=f"Falha ao criar pagamento no Stripe: {exc}"
            )

    if payment_intent_id is None:
        payment_intent_id = f"pi_demo_{uuid.uuid4().hex[:24]}"
        client_secret = f"{payment_intent_id}_secret_{uuid.uuid4().hex[:16]}"
        order.stripe_payment_intent = payment_intent_id

    db.commit()
    db.refresh(order)

    event = tickets[0].event
    return schemas.OrderResponse(
        order_id=order.id,
        total_amount=total_amount,
        payment_intent_id=payment_intent_id,
        client_secret=client_secret,
        status=order.status.value,
        created_at=order.created_at,
        event_title=event.title if event else None,
        seats=[t.seat_number for t in tickets],
        demo_mode=demo_mode,
    )


def get_order(db: Session, order_id: str, user: models.User) -> schemas.OrderDetailResponse:
    order = db.query(models.Order).filter(models.Order.id == order_id).first()
    if order is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Pedido não encontrado: {order_id}")
    if order.user_id != user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Este pedido não pertence a este usuário")
    return _map_detail(db, order)


def get_my_orders(db: Session, user: models.User) -> List[schemas.OrderDetailResponse]:
    orders = (
        db.query(models.Order)
        .filter(models.Order.user_id == user.id)
        .order_by(models.Order.created_at.desc())
        .all()
    )
    return [_map_detail(db, o) for o in orders]


async def confirm_payment(db: Session, order_id: str, user: models.User) -> schemas.OrderDetailResponse:
    """Confirma o pagamento (fluxo demo, ou chamado pelo webhook do Stripe)."""
    order = db.query(models.Order).filter(models.Order.id == order_id).first()
    if order is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pedido não encontrado")
    if order.user_id != user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Este pedido não pertence a este usuário")
    if order.status == models.OrderStatus.PAID:
        return _map_detail(db, order)
    if order.status != models.OrderStatus.PENDING:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Pedido não pode mais ser pago")

    await process_successful_payment(db, order.stripe_payment_intent)
    db.refresh(order)
    return _map_detail(db, order)


async def process_successful_payment(db: Session, payment_intent_id: str) -> None:
    order = db.query(models.Order).filter(models.Order.stripe_payment_intent == payment_intent_id).first()
    if order is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pedido não encontrado para este pagamento")

    order.status = models.OrderStatus.PAID
    db.commit()

    for item in order.order_items:
        await ticket_service.confirm_ticket_sale(db, item.ticket_id, order.id)


def _map_detail(db: Session, order: models.Order) -> schemas.OrderDetailResponse:
    items = []
    for oi in order.order_items:
        ticket = oi.ticket
        event = ticket.event if ticket else None
        qr = None
        if order.status == models.OrderStatus.PAID and ticket and event:
            qr = generate_qr_code_base64(
                f"EVENTPLATFORM|order={order.id}|seat={ticket.seat_number}|event={event.title}"
            )
        items.append(
            schemas.OrderItemTicket(
                seat_number=ticket.seat_number if ticket else "?",
                price=oi.price,
                qr_code_base64=qr,
                event_title=event.title if event else None,
                event_date=event.date if event else None,
                event_location=event.location if event else None,
            )
        )

    event_title = items[0].event_title if items else None
    return schemas.OrderDetailResponse(
        order_id=order.id,
        total_amount=order.total_amount,
        payment_intent_id=order.stripe_payment_intent,
        status=order.status.value,
        created_at=order.created_at,
        event_title=event_title,
        seats=[i.seat_number for i in items],
        demo_mode=_stripe is None,
        items=items,
    )

File: src/services/orderService.js
'use strict';
const crypto = require('crypto');
const OrderModel = require('../models/Order');
const OrderItemModel = require('../models/OrderItem');
const TicketModel = require('../models/Ticket');
const EventModel = require('../models/Event');
const { TicketStatus, OrderStatus } = require('../models/enums');
const { HttpError } = require('../utils/errors');
const { generateQrCodeBase64 } = require('../utils/qr');
const ticketService = require('./ticketService');

const STRIPE_SECRET_KEY = (process.env.STRIPE_SECRET_KEY || '').trim();

let _stripe = null;
if (STRIPE_SECRET_KEY) {
  try {
    // Dependência opcional: só é necessária se STRIPE_SECRET_KEY estiver configurada.
    const Stripe = require('stripe');
    _stripe = new Stripe(STRIPE_SECRET_KEY);
  } catch (err) {
    _stripe = null;
  }
}

function shortId() {
  return crypto.randomBytes(12).toString('hex');
}

async function createOrder(data, user) {
  const tickets = [];
  let totalAmount = 0;

  for (const ticketId of data.ticket_ids) {
    const ticket = TicketModel.findById(ticketId);
    if (!ticket) {
      throw new HttpError(404, `Ingresso não encontrado: ${ticketId}`);
    }
    if (ticket.status !== TicketStatus.RESERVED || ticket.reserved_by !== user.id) {
      throw new HttpError(
        409,
        `Ingresso não está reservado para este usuário: ${ticketId}`
      );
    }
    tickets.push(ticket);
    totalAmount += Number(ticket.price);
  }

  let order = OrderModel.create({
    userId: user.id,
    totalAmount,
    status: OrderStatus.PENDING,
  });

  for (const ticket of tickets) {
    OrderItemModel.create({ orderId: order.id, ticketId: ticket.id, price: ticket.price });
  }

  let paymentIntentId = null;
  let clientSecret = null;
  let demoMode = true;

  if (_stripe !== null) {
    try {
      const intent = await _stripe.paymentIntents.create({
        amount: Math.round(totalAmount * 100),
        currency: 'usd',
        metadata: { orderId: order.id, userId: user.id },
      });
      paymentIntentId = intent.id;
      clientSecret = intent.client_secret;
      demoMode = false;
    } catch (exc) {
      throw new HttpError(502, `Falha ao criar pagamento no Stripe: ${exc.message || exc}`);
    }
  }

  if (paymentIntentId === null) {
    paymentIntentId = `pi_demo_${shortId()}`;
    clientSecret = `${paymentIntentId}_secret_${shortId().slice(0, 16)}`;
  }
  order = OrderModel.setStripePaymentIntent(order.id, paymentIntentId);

  const event = EventModel.findById(tickets[0].event_id);

  return {
    order_id: order.id,
    total_amount: totalAmount,
    payment_intent_id: paymentIntentId,
    client_secret: clientSecret,
    status: order.status,
    created_at: order.created_at,
    event_title: event ? event.title : null,
    seats: tickets.map((t) => t.seat_number),
    demo_mode: demoMode,
  };
}

function getOrder(orderId, user) {
  const order = OrderModel.findById(orderId);
  if (!order) {
    throw new HttpError(404, `Pedido não encontrado: ${orderId}`);
  }
  if (order.user_id !== user.id) {
    throw new HttpError(403, 'Este pedido não pertence a este usuário');
  }
  return mapDetail(order);
}

function getMyOrders(user) {
  return OrderModel.findByUser(user.id).map((o) => mapDetail(o));
}

async function confirmPayment(orderId, user) {
  const order = OrderModel.findById(orderId);
  if (!order) {
    throw new HttpError(404, 'Pedido não encontrado');
  }
  if (order.user_id !== user.id) {
    throw new HttpError(403, 'Este pedido não pertence a este usuário');
  }
  if (order.status === OrderStatus.PAID) {
    return mapDetail(order);
  }
  if (order.status !== OrderStatus.PENDING) {
    throw new HttpError(409, 'Pedido não pode mais ser pago');
  }

  await processSuccessfulPayment(order.stripe_payment_intent);
  const refreshed = OrderModel.findById(orderId);
  return mapDetail(refreshed);
}

async function processSuccessfulPayment(paymentIntentId) {
  const order = OrderModel.findByPaymentIntent(paymentIntentId);
  if (!order) {
    throw new HttpError(404, 'Pedido não encontrado para este pagamento');
  }

  OrderModel.setStatus(order.id, OrderStatus.PAID);

  const items = OrderItemModel.findByOrder(order.id);
  for (const item of items) {
    await ticketService.confirmTicketSale(item.ticket_id, order.id);
  }
}

function mapDetail(order) {
  const orderItems = OrderItemModel.findByOrder(order.id);
  const items = orderItems.map((oi) => {
    const ticket = TicketModel.findById(oi.ticket_id);
    const event = ticket ? EventModel.findById(ticket.event_id) : null;
    return {
      _ticket: ticket,
      _event: event,
      seat_number: ticket ? ticket.seat_number : '?',
      price: Number(oi.price),
      event_title: event ? event.title : null,
      event_date: event ? event.date : null,
      event_location: event ? event.location : null,
    };
  });

  const eventTitle = items.length > 0 ? items[0].event_title : null;

  return {
    order_id: order.id,
    total_amount: Number(order.total_amount),
    payment_intent_id: order.stripe_payment_intent,
    status: order.status,
    created_at: order.created_at,
    event_title: eventTitle,
    seats: items.map((i) => i.seat_number),
    demo_mode: _stripe === null,
    items: items.map(({ _ticket, _event, ...rest }) => rest),
  };
}

/**
 * Gera QR codes para os itens de um pedido pago. Feito de forma assíncrona
 * e separada de mapDetail (que é síncrono) porque a geração de QR code é
 * assíncrona nesta implementação em Node.
 */
async function attachQrCodes(detail) {
  if (detail.status !== OrderStatus.PAID) {
    return detail;
  }
  const itemsWithQr = [];
  for (const item of detail.items) {
    const qr = await generateQrCodeBase64(
      `EVENTPLATFORM|order=${detail.order_id}|seat=${item.seat_number}|event=${item.event_title}`
    );
    itemsWithQr.push({ ...item, qr_code_base64: qr });
  }
  return { ...detail, items: itemsWithQr };
}

module.exports = {
  createOrder,
  getOrder,
  getMyOrders,
  confirmPayment,
  processSuccessfulPayment,
  attachQrCodes,
};
